### PR TITLE
Feature/rocky9 signalbackground pr

### PR DESCRIPTION
**Pull request by MattOsbiston**
A low energy reconstruction capable of handling radiological backgrounds, designed for neutrino interactions of 5 to 30 MeV, currently tested using samples using the CC Nue and e-ES interaction channels. Updates include compatibility with Dom's update to larpandora, introducing a nuance code == 4000 for nu interaction of SN origin

### CODE REVIEW

File: larpandoracontent/LArHelpers/LArHierarchyHelper.cc
Comment (MattOsbiston) [inline larpandoracontent/LArHelpers/LArHierarchyHelper.cc:1]: These changes many no longer be required following PR #225 - I had to remove some features in order for the algorithm to run when no PFO/no Vertex found, but may be fine now.

File: larpandoracontent/LArHelpers/LArMCParticleHelper.cc
Comment (MattOsbiston) [inline larpandoracontent/LArHelpers/LArMCParticleHelper.cc:1]: Changes here seem unnecessary

File: larpandoracontent/LArMonitoring/HierarchyValidationAlgorithm.cc
Comment (MattOsbiston) [inline larpandoracontent/LArMonitoring/HierarchyValidationAlgorithm.cc:1]: Only important changes here are adding XML parameters for min Purity, Completeness, Hits and Good View hits. Total energy could be a useful tree variable for others, and I use it for latest work too - so perhaps these changes could be added in to John B's PR

File: larpandoracontent/LArThreeDReco/LArPfoRecovery/VertexBasedPfoRecoveryAlgorithm.cc
Comment (MattOsbiston) [inline larpandoracontent/LArThreeDReco/LArPfoRecovery/VertexBasedPfoRecoveryAlgorithm.cc:1]: All changes here seem unnecessary

File: larpandoracontent/LArThreeDReco/LArTransverseTrackMatching/ClearTracksTool.cc
Comment (MattOsbiston) [inline larpandoracontent/LArThreeDReco/LArTransverseTrackMatching/ClearTracksTool.cc:1]: Not a change from me, from rebase?

**Review by MattOsbiston — COMMENTED** [review]
The bulk of the review is looking over the two new additional algorithms: DLSignal Algorithm and MVAClusterMergingAlgorithm. Besides that, the other changes are main for XML configurable for lower thresholds for low e compatability.

File: larpandoracontent/LArThreeDReco/LArPfoRecovery/VertexBasedPfoRecoveryAlgorithm.cc
Comment (AndyChappell) [inline larpandoracontent/LArThreeDReco/LArPfoRecovery/VertexBasedPfoRecoveryAlgorithm.cc:1]: I'm not convinced of a good physical motivation for the lowe fit pitch change here. More generally, there are a lot of formatting changes in this file that probably shouldn't be present. It looks to me like the only intentional change is the fit pitch, so this file should be reverted, and then add in only the conditional fit pitch change, and then we should separately discuss why this value is motivated.

File: larpandoracontent/LArThreeDReco/LArTransverseTrackMatching/ClearTracksTool.cc
Comment (AndyChappell) [inline larpandoracontent/LArThreeDReco/LArTransverseTrackMatching/ClearTracksTool.cc:1]: Revert this file.

File: larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+			      if ( distance < m_proximityThreshold)
+			      { 
+				      ++proximity;
+				      if ( distance < m_contactThreshold)
+				      {
+					      ++contact;
+					      --proximity;
+				      }
+		               }
+
```
Comment (AndyChappell) [inline larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc:326]: Whitespace

File: larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+    return clusterMC;
+}
+
+//---------------------------------------------------------------------------------------------//
+
+template <typename T>
+bool MvaLowEClusterMergingAlgorithm<T>::IsValidToUse(const Cluster *const cluster, std::map<const Cluster*, bool> &clusterIsUsed) const
+{
+
```
Comment (AndyChappell) [inline larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc:142]: Whitespace

File: larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+        catch (StatusCodeException &statusCodeException)
+        {
+            throw statusCodeException;
+        }
+        
+	if (m_writeTree)
+        {
+            PANDORA_MONITORING_API(SaveTree(this->GetPandora(), m_treeName.c_str(), m_fileName.c_str(), "RECREATE"))
+         }
+
```
Comment (AndyChappell) [inline larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc:102]: Whitespace

File: larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+//------------------------------------------------------------------------------------------------------------------------------------------
+template <typename T>
+MvaLowEClusterMergingAlgorithm<T>::~MvaLowEClusterMergingAlgorithm()
+{
+    if (m_writeTree)
+        {
+            PANDORA_MONITORING_API(SaveTree(this->GetPandora(), m_treeName.c_str(), m_fileName.c_str(), "RECREATE"));
+         }
+
```
Comment (AndyChappell) [inline larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc:64]: Whitespace

File: larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+template <typename T>
+MvaLowEClusterMergingAlgorithm<T>::~MvaLowEClusterMergingAlgorithm()
+{
+    if (m_writeTree)
+        {
+            PANDORA_MONITORING_API(SaveTree(this->GetPandora(), m_treeName.c_str(), m_fileName.c_str(), "RECREATE"));
+         }
```
Comment (AndyChappell) [inline larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc:63]: clang format can handle the indentation issues

File: larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+    m_upperHitThreshold{100},
+    m_filePathEnvironmentVariable{"FW_SEARCH_PATH"},
+    m_countHitsThreshold{0},
+    m_vertexListName{"NeutrinoVertices3D"},
+    m_contactThreshold{2},
+    m_proximityThreshold{5},
+    m_divisions{8},
+    m_sectorTolerance{0.03},
+    m_printOut{false}
+
```
Comment (AndyChappell) [inline larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc:52]: Whitespace

File: larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    float                  m_countHitsThreshold;    ///< A cut on whether cluster merges will occur depending on total event hits
+    std::string            m_vertexListName;        ///< Input Vertex List name for vertex based calculation
+    float                  m_contactThreshold;      ///< Distance value for hits to be considered in contact
+    float                  m_proximityThreshold;    ///< Distance value for hits to be considered in proximity   
+    float                  m_divisions;             ///< Number of sectors to search in with Edge Hit Finder Function
+    float                  m_sectorTolerance;       ///< Tolerance in radians for dot product between sector and centroid to CaloHit vector
+    bool                   m_printOut;              ///< Whether the algorithm outputs cluster size information
+
+
+
```
Comment (AndyChappell) [inline larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.h:88]: Whitespace

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.h [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+     *
+     *  @return whether the clusters are associated
+     */
+    bool AreClustersAssociated(const pandora::CartesianVector &innerClusterEnd, const pandora::CartesianVector &outerClusterStart,
+        const pandora::ClusterFitResult &innerFit, const pandora::ClusterFitResult &outerFit) const;
+
+    unsigned int m_minClusterLayers;     ///< minimum allowed number of layers for a clean cluster
+    unsigned int m_maxGapLayers;         ///< maximum allowed number of layers between associated clusters
+    unsigned int m_fitLayers;            ///< number of layers to fit at start and end of cluster
+    float m_maxGapDistanceSquared;       ///< maximum allowed distance (squared) between associated clusters
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.h:66]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.h [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+     *  @param  innerFit inner cluster fit result
+     *  @param  outerFit outer cluster fit result
+     *
+     *  @return whether the clusters are associated
+     */
+    bool AreClustersAssociated(const pandora::CartesianVector &innerClusterEnd, const pandora::CartesianVector &outerClusterStart,
+        const pandora::ClusterFitResult &innerFit, const pandora::ClusterFitResult &outerFit) const;
+
+    unsigned int m_minClusterLayers;     ///< minimum allowed number of layers for a clean cluster
+    unsigned int m_maxGapLayers;         ///< maximum allowed number of layers between associated clusters
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.h:64]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    m_hitSizeX(0.5f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void CheatingLongitudinalAssociationAlgorithm::GetListOfCleanClusters(const ClusterList *const pClusterList, ClusterVector &clusterVector) const
+{
+    for (ClusterList::const_iterator iter = pClusterList->begin(), iterEnd = pClusterList->end(); iter != iterEnd; ++iter)
+    {
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc:39]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        const Cluster *const pCluster = *iter;
+
+        if (1 + pCluster->GetOuterPseudoLayer() - pCluster->GetInnerPseudoLayer() < m_minClusterLayers)
+            continue;
+
+        clusterVector.push_back(pCluster);
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc:45]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+void CheatingLongitudinalAssociationAlgorithm::PopulateClusterAssociationMap(const ClusterVector &clusterVector, ClusterAssociationMap &clusterAssociationMap) const
+{
+    // ATTN This method assumes that clusters have been sorted by layer
+    for (ClusterVector::const_iterator it1 = clusterVector.begin(), it1End = clusterVector.end(); it1 != it1End; ++it1)
+    {
+        const Cluster *const pInnerCluster = *it1;
+
+        for (ClusterVector::const_iterator it2 = it1, it2End = clusterVector.end(); it2 != it2End; ++it2)
+        {
+            const Cluster *const pOuterCluster = *it2;
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc:62]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------------------------------------------------------------------
+StatusCode CheatingLongitudinalAssociationAlgorithm::GetPrincipalAxisVector() const
+{ 
+    const ClusterList *pClusterList(nullptr);
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS,!=,PandoraContentApi::GetCurrentList(*this,pClusterList));
+    const std::string m_treename{"mc"};
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc:81]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+        if(!eigenVecs.empty())
+        {
+            const float pcaX{eigenVecs.front().GetX()};
+            const float pcaY{eigenVecs.front().GetY()};
+            const float pcaZ{eigenVecs.front().GetZ()};	
+ 
+            for(const CartesianVector &axis : eigenVecs)	
+       	    {
+                std::cout << axis  << std::endl;
+            }
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc:102]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool CheatingLongitudinalAssociationAlgorithm::AreClustersAssociated(const Cluster *const pInnerCluster, const Cluster *const pOuterCluster) const
+{
+    const MCParticle *const pMCOuterParticle(MCParticleHelper::GetMainMCParticle(pOuterCluster));
+    const MCParticle *const pMCInnerParticle(MCParticleHelper::GetMainMCParticle(pInnerCluster)); 
+    std::cout << "Cluster, Inner: ";
+    std::cout << pMCInnerParticle->GetParticleId() << std::endl;
+    std::cout << "Cluster, Outer: " << pMCOuterParticle->GetParticleId() << std::endl;
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc:138]: Can be reverted

File: larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+     
+    if (pMCOuterParticle == pMCInnerParticle)
+        return true;
+
+    return false;
+}
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingLongitudinalAssociationAlgorithm.cc:144]: Can be reverted

File: larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc [hunk 17]
```diff
@@ -130,26 +132,35 @@ void CheatingPfoCreationAlgorithm::CreatePfos(const MCParticleToClusterListMap &
         try
         {
             PandoraContentApi::ParticleFlowObject::Parameters pfoParameters;
-            pfoParameters.m_particleId = (this->IsShower(pMCParticle) ? E_MINUS : MU_MINUS);
+            pfoParameters.m_particleId = pMCParticle->GetParticleId();
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc:135]: Can be reverted

File: larpandoracontent/LArContent.cc [hunk 18]
```diff
@@ -69,13 +71,6 @@
 
 #include "larpandoracontent/LArPlugins/LArParticleIdPlugins.h"
 
-#include "larpandoracontent/LArShowerRefinement/ConnectionPathwayFeatureTool.h"
-#include "larpandoracontent/LArShowerRefinement/ElectronInitialRegionRefinementAlgorithm.h"
-#include "larpandoracontent/LArShowerRefinement/PeakDirectionFinderTool.h"
-#include "larpandoracontent/LArShowerRefinement/ProtoShowerMatchingTool.h"
-#include "larpandoracontent/LArShowerRefinement/ShowerSpineFinderTool.h"
-#include "larpandoracontent/LArShowerRefinement/ShowerStartFinderTool.h"
-
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArContent.cc:78]: Can be reverted

File: larpandoracontent/LArContent.cc [hunk 19]
```diff
@@ -225,7 +221,6 @@
 // clang-format off
 #define LAR_ALGORITHM_LIST(d)                                                                                                   \
     d("LArMuonLeadingEventValidation",          MuonLeadingEventValidationAlgorithm)                                            \
-    d("LArElectronInitialRegionRefinement",     ElectronInitialRegionRefinementAlgorithm)                                       \
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArContent.cc:228]: Can be reverted

File: larpandoracontent/LArContent.cc [hunk 20]
```diff
@@ -337,14 +335,6 @@
...
 
 #define LAR_ALGORITHM_TOOL_LIST(d)                                                                                              \
-    d("LArConnectionRegionFeatureTool",         ConnectionRegionFeatureTool)                                                    \
-    d("LArShowerRegionFeatureTool",             ShowerRegionFeatureTool)                                                        \
-    d("LArAmbiguousRegionFeatureTool",          AmbiguousRegionFeatureTool)                                                     \
-    d("LArInitialRegionFeatureTool",            InitialRegionFeatureTool)                                                       \
-    d("LArPeakDirectionFinder",                 PeakDirectionFinderTool)                                                        \
-    d("LArProtoShowerMatching",                 ProtoShowerMatchingTool)                                                        \
-    d("LArShowerSpineFinder",                   ShowerSpineFinderTool)                                                          \
-    d("LArShowerStartFinder",                   ShowerStartFinderTool)                                                          \
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArContent.cc:347]: Can be reverted

File: larpandoracontent/LArHelpers/LArMCParticleHelper.cc [hunk 21]
```diff
@@ -29,8 +29,8 @@ namespace lar_content
 using namespace pandora;
 
 LArMCParticleHelper::PrimaryParameters::PrimaryParameters() :
-    m_minPrimaryGoodHits(15),
-    m_minHitsForGoodView(5),
+    m_minPrimaryGoodHits(30),
+    m_minHitsForGoodView(10),
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArHelpers/LArMCParticleHelper.cc:33]: Can be reverted

File: larpandoracontent/LArHelpers/LArMCParticleHelper.h [hunk 22]
```diff
@@ -110,7 +110,7 @@ class LArMCParticleHelper
      */
     static bool IsCosmicRay(const pandora::MCParticle *const pMCParticle);
 
-    /**
+     /**
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArHelpers/LArMCParticleHelper.h:113]: Can be reverted

File: larpandoracontent/LArMonitoring/HierarchyMonitoringAlgorithm.cc [hunk 23]
```diff
@@ -216,6 +216,7 @@ void HierarchyMonitoringAlgorithm::VisualizeMCProcess(const LArHierarchyHelper::
         {"ioni", 5}, {"brem", 6}, {"conv", 3}, {"capture", 6}, {"inelastic", 9}, {"elastic", 8}, {"decay", 7}, {"coulomb", 9},
         {"pair_prod", 4}, {"transport", 1}, {"rayleigh", 9}, {"kill", 2}, {"nuclear", 5}, {"background", 7}};
 
+
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArMonitoring/HierarchyMonitoringAlgorithm.cc:219]: Can be reverted

File: larpandoracontent/LArMonitoring/MCParticleMonitoringAlgorithm.cc [hunk 24]
```diff
@@ -23,9 +23,12 @@ using namespace pandora;
 namespace lar_content
 {
 
-MCParticleMonitoringAlgorithm::MCParticleMonitoringAlgorithm() :
-    m_useTrueNeutrinosOnly(false),
-    m_minHitsForDisplay(1)
+
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArMonitoring/MCParticleMonitoringAlgorithm.cc:26]: Can be reverted

File: larpandoracontent/LArMonitoring/MCParticleMonitoringAlgorithm.h [hunk 25]
```diff
@@ -52,6 +52,8 @@ class MCParticleMonitoringAlgorithm : public pandora::Algorithm
 
     bool m_useTrueNeutrinosOnly;      ///< Whether to consider only mc particles that were neutrino induced
     unsigned int m_minHitsForDisplay; ///< Min hits associated with mc particle to warrant display to terminal
+    unsigned int m_minPrimaryGoodHits;         ///< Minimum hits which are deemed useful
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArMonitoring/MCParticleMonitoringAlgorithm.h:55]: Can be reverted

File: larpandoracontent/LArMonitoring/TestBeamEventValidationAlgorithm.cc [hunk 26]
```diff
@@ -127,7 +127,7 @@ void TestBeamEventValidationAlgorithm::ProcessOutput(
     MCParticleList associatedMCPrimaries;
 
     int nCorrectTB(0), nTotalTB(0), nCorrectCR(0), nTotalCR(0), nFakeTB(0), nFakeCR(0), nSplitTB(0), nSplitCR(0), nLost(0);
-    int mcPrimaryIndex(0), nTargetMatches(0), nTargetTBMatches(0), nTargetCRMatches(0);
+    int mcPrimaryIndex(0), nTargetMatches(0), nTargetTBMatches(0), nTargetCRMatches(0), nTargetGoodTBMatches(0);
```
[the remark this answers is not in the document]
Comment (MattOsbiston) [reply larpandoracontent/LArMonitoring/TestBeamEventValidationAlgorithm.cc:130]: Can be reverted

**Comment by AndyChappell** [comment]
Will be superseded by a rebased version

**Pull request closed without merging**